server/middleware_gzip.go

```diff
@@ -0,0 +1,90 @@
+package server
+
+import (
+	"bytes"
+	"compress/gzip"
+	"io"
+	"net/http"
+	"strings"
+)
+
+// GZipHandler is responsible for determining if the incoming request should be served gzipped data.
+// All response data is assumed to be compressed prior to being passed to this handler.
+//
+// If the incoming request has the "Accept-Encoding" header set with the values of "gzip" or "*"
+// the response header "Content-Encoding: gzip" is set and the compressed data is returned.
+//
+// If no "Accept-Encoding" header is present or "Accept-Encoding" has a value of "gzip;q=0" or
+// "*;q=0" the response is decompressed prior to being sent to the client.
+func GZipHandler(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+
+		acceptEncoding := r.Header.Get("Accept-Encoding")
+		if acceptEncoding == "" {
+			// decompress
+			next.ServeHTTP(&gzipDecompressResponseWriter{resp: w}, r)
+			return
+		}
+
+		for _, v := range strings.Split(acceptEncoding, ",") {
+			v = strings.TrimSpace(strings.ToLower(v))
+			if strings.Contains(v, "gzip") || strings.Contains(v, "*") {
+				if strings.HasSuffix(v, ";q=0") {
+					//	decompress
+					next.ServeHTTP(&gzipDecompressResponseWriter{resp: w}, r)
+					return
+				}
+				break
+			} else {
+				//	decompress
+				next.ServeHTTP(&gzipDecompressResponseWriter{resp: w}, r)
+				return
+			}
+		}
```

Comment:
```go
decompress := false
for _, v := range strings.Split(acceptEncoding, ",") {
         if  (strings.Contains(v,"gzip") || strings.Contains(v,"*")) && !strings.HasSuffix(v,";q=0") {
           decompress = true
        } 
}
if decompress {
        next.ServeHTTP(&gzipDecompressResponseWriter{resp: w}, r)
	return
}
// set appropriate header
w.Header().Set("Content-Encoding", "gzip")
next.ServeHTTP(w, r)
return
```

This is cleaner to read and fixes the issue if the gzip header is the second one. Please, add a test case for this.